src/citrine/informatics/design_spaces/hierarchical_design_space.py

```diff
@@ -0,0 +1,215 @@
+
+    Material histories produced by this design space are connected based on the
+    name identifiers and formulation contents of each node.
+    For example, the root node may contain a formulation subspace with an ingredient named
+    'New Mixture-001'. If this ingredient matches the `name` field on one of the sub-nodes
+    in the design space, the resulting candidates will contain a terminal material with this
+    new mixture as one of its ingredients. This procedure can be extended to sub-nodes
+    referencing other sub-nodes, allowing for the linkage of complex material history shapes
+    in the resulting candidates.
+
+    Every node also contains a set of :class:`~citrine.informatics.dimensions.Dimension`s
+    used to define any attributes (i.e., properties, processing parameters)
+    that will appear on the materials produced by that node.
+
+    :class:`~citrine.informatics.data_sources.DataSource`s can be included on the configuration
+    to allow for design over "known" materials. The Citrine Platform will look up
+    the any ingredient names from formulation subspaces on the design space nodes
+    in order to inject their composition/properties into the material history of the candidates.
+    When constructing a default hierarchical design space,
+    the Citrine Platform includes any data sources found on the provided predictor configuration.
+
+    Parameters
+    ----------
+    name: str
+        Name of the design space
+    description: str
+        Description of the design space
+    root: MaterialNodeDefinition
+        Terminal material node produced by the design space
+    subspaces: List[MaterialNodeDefinition]
+        Sub material nodes produced by the design space
+    data_sources: List[DataSource]
+        Data sources used to inject known materials into the hierarchical candidates
+
+    """
+
+    root = properties.Object(MaterialNodeDefinition, "data.instance.root")
+    subspaces = properties.List(
+        properties.Object(MaterialNodeDefinition), "data.instance.subspaces", default=list()
+    )
+    data_sources = properties.List(
+        properties.Object(DataSource), "data.instance.data_sources", default=list()
+    )
+    typ = properties.String(
+        "data.instance.type", default="HierarchicalDesignSpace", deserializable=False
+    )
+
+    def __init__(
+            self,
+            name: str,
+            *,
+            description: str,
+            root: MaterialNodeDefinition,
+            subspaces: Optional[List[MaterialNodeDefinition]] = None,
+            data_sources: Optional[List[DataSource]] = None
+    ):
+        self.name: str = name
+        self.description: str = description
+        self.root: MaterialNodeDefinition = root
+        self.subspaces: List[MaterialNodeDefinition] = subspaces or list()
+        self.data_sources: List[DataSource] = data_sources or list()
+
+    def _post_dump(self, data: dict) -> dict:
```

Comment:
All of this pre-build and post-dump stuff is necessary because the `MaterialNodeDefinition` class contains a `FormulationDesignSpace`. Because of that, we have to wrap that JSON in the `DesignSpace.wrap_instance` format such that it can be built properly.